docs/advanced/additional-requests.rst

```diff
@@ -750,9 +750,12 @@ doc section) explicitly raises when implementing it for **web-poet** should be
 For frameworks that implement and use **web-poet**, exceptions that ocurred when
 handling the additional requests like `connection errors`, `time outs`, `TLS
 errors`, etc should be replaced by :class:`web_poet.exceptions.http.HttpRequestError`
-by raising it explicitly. For responses that are successful but don't have a ``200``
-**status code**, this exception shouldn't be raised at all. Instead, the
-:class:`~.HttpResponse` should simply reflect the response contents as is.
+by raising it explicitly.
+
+For responses that are not really errors like in the 100-3xx status code range,
+this exception shouldn't be raised at all. However, for responses with status
+codes in the 400-5xx range, the implementing framework should properly raise
+the exception.
```

Comment:
Hey! What do you think about a slightly different approach?

1. State that frameworks must not raise exceptions for 400-5xx status code. They should return an HttpResponse, with a proper status code set. I'm actually not sure how allow_status works in this PR, if a framework raises an exception for 400-5xx codes.
2. Frameworks must raise HttpRequestError for connections errors, etc. I'm not sure about timeouts though; it's fine to make frameworks handle it for now, but we'd need to think about it further.
3. We should define framework behavior for 3xx status codes, for redirects. For example, we can say that frameworks must follow redirects.
4. Probably it's a right time to add another exception class, to be able to distinguish errors without a response, and errors where response is present. We can even have a response as an attribute of this exception: because of (1), this information would be available.

Reply:
Hi @kmike , let me know what do you think about the points below?

1. Sounds good. It's currently the expectation on the `master` branch. So not much changes here aside from perhaps improving the docs to reflect on it more clearly.

    - I was also thinking of perhaps having a similar interface to [aiohttp](https://docs.aiohttp.org/en/stable/client_reference.html?highlight=errors#aiohttp.ClientResponse.raise_for_status) and [request](https://2.python-requests.org/en/master/api/#requests.Response.raise_for_status) `raise_for_status()` method could cover 400-5xx errors. This shifts the focus of having the errors being the default into an intentional one dictated by the developer.

2. We'll need to improve the docs on this regard regarding raising `HttpRequestError` for connection errors. For timeouts, we can **create** a new exception subclassing from `HttpRequestError`.

3. Got it, we'll need to improve the docs for this one. Perhaps **create** an entirely new section for a framework's checklist.

4. I think we can use the current `HttpRequestError` for errors without a response and **create** a new `HttpReponseError` for errors with responses. In addition, we could also **create** a parent class for both of them, perhaps calling it `HttpError`.

We can create a new PR for this as it has a different approach due to **(1)**.

Reply:
Hey @BurnzZ! Regarding (1) - I was thinking that frameworks would not be raising these exceptions, but web-poet would.

Reply:
I see, I got that wrong. Thanks for the clarification! Will edit my points above.

Reply:
Regarding exceptions: +1 to have 3 exceptions - a base exception, and 2 exceptions for these conditions.

Reply:
As for timeouts, the main issue is who sets the timeout value - is it a framework, or is it something which can be controlled in PO? I think so far we've been side-stepping it, and letting framework to manage it. 

But it means that  using a specific timeout for a specific request, e.g. 
* decreasing it for requests which should be fast, or 
* increasing for requests which are expected to take long, or 
* setting a timeout if a library doesn't have a timeout

can't be done. I don't think this is something to solve in this PR, but it's something to think about.

Reply:
Hi @kmike ! I've reorganized the docs to better document all of these things. The main change is to properly distinguish the docs for Page Object developers vs developers that would implement a framework for **web-poet**. https://github.com/scrapinghub/web-poet/pull/38/commits/3ebb2973aaaea79bff7b9f4f033628233b9d8b87.

The new base class exception named `HttpError` was also introduced. https://github.com/scrapinghub/web-poet/pull/38/commits/470bc6f417431168d0138762dbf76d12801163c4.

Kindly let me know if there are other specific things we need to explicitly document here that were missed out.

----

Regarding the timeouts, yeah, that's indeed tricky. It also adds some additional complexity to Page Objects, since aside from knowing how to construct the requests, it would also need to consider _(and establish an expectation)_ how long it would take the response to arrive.

There would be some factors that the Page Objects might not have access to, like the number of requests made to the same domain in the past minute _(which could timeout due to rate limiting)_ or the shifting tide of non-`200` responses for a given domain _(where the site could throttle down requests as well)_. Letting a Page Object decide on timeouts based on limited knowledge is difficult.

I think at least for the short-term, we can let the framework handle it.